src/secp256k1.c

```diff
@@ -588,10 +589,189 @@ int secp256k1_ec_pubkey_combine(const secp256k1_context* ctx, secp256k1_pubkey *
+    unsigned char tweak[32];
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(commitment != NULL);
+    ARG_CHECK(pubkey != NULL);
+    ARG_CHECK(data != NULL);
+
+    *commitment = *pubkey;
+    if (!secp256k1_ec_commit_tweak(ctx, tweak, commitment, data, data_size)) {
+        return 0;
+    }
+    return secp256k1_ec_pubkey_tweak_add(ctx, commitment, tweak);
+}
+
+/* Compute the seckey of an ec commitment from the original secret key of the pubkey as seckey +
+ * hash(pubkey, data). */
+static int secp256k1_ec_commit_seckey(const secp256k1_context* ctx, unsigned char *seckey, const secp256k1_pubkey *pubkey, const unsigned char *data, size_t data_size) {
+    unsigned char tweak[32];
+    secp256k1_pubkey pubkey_tmp;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(seckey != NULL);
+    ARG_CHECK(data != NULL);
+
+    if (pubkey == NULL) {
+        /* Compute pubkey from seckey if not provided */
+        int overflow;
+        secp256k1_scalar x;
+        secp256k1_gej pj;
+        secp256k1_ge p;
+
+        secp256k1_scalar_set_b32(&x, seckey, &overflow);
+        if (overflow != 0) {
+            return 0;
+        }
+        secp256k1_ecmult_gen(&ctx->ecmult_gen_ctx, &pj, &x);
+        secp256k1_ge_set_gej(&p, &pj);
+        secp256k1_pubkey_save(&pubkey_tmp, &p);
+        pubkey = &pubkey_tmp;
+    }
+
+    if (!secp256k1_ec_commit_tweak(ctx, tweak, pubkey, data, data_size)) {
+        return 0;
+    }
+    return secp256k1_ec_privkey_tweak_add(ctx, seckey, tweak);
+}
+
+/* Verify an ec commitment as pubkey + hash(pubkey, data)*G ?= commitment. */
+static int secp256k1_ec_commit_verify(const secp256k1_context* ctx, const secp256k1_pubkey *commitment, const secp256k1_pubkey *pubkey, const unsigned char *data, size_t data_size) {
+    secp256k1_gej pj;
+    secp256k1_ge p;
+    secp256k1_pubkey commitment_tmp;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(commitment != NULL);
+    ARG_CHECK(pubkey != NULL);
+    ARG_CHECK(data != NULL);
+
+    if (!secp256k1_ec_commit(ctx, &commitment_tmp, pubkey, data, data_size)) {
+        return 0;
+    }
+
+    /* Return commitment == commitment_tmp */
+    secp256k1_gej_set_infinity(&pj);
+    secp256k1_pubkey_load(ctx, &p, &commitment_tmp);
+    secp256k1_gej_add_ge(&pj, &pj, &p);
+    secp256k1_pubkey_load(ctx, &p, commitment);
+    secp256k1_ge_neg(&p, &p);
+    secp256k1_gej_add_ge(&pj, &pj, &p);
+    return secp256k1_gej_is_infinity(&pj);
+}
+
+static uint64_t s2c_commit_context_magic = 0xd5bafd089f7e1c63;
+int secp256k1_s2c_commit_context_create(secp256k1_context *ctx, secp256k1_s2c_commit_context *s2c_ctx, const unsigned char *data32) {
+    secp256k1_sha256 sha;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(s2c_ctx != NULL);
+    ARG_CHECK(data32 != NULL);
+
+    memcpy(s2c_ctx->magic, &s2c_commit_context_magic, sizeof(s2c_ctx->magic));
```

Comment:
Probably me just not familiar enough with C but why copy from a `uint64_t` to a `unsigned char[8]` instead of just putting a `uint64_t` in the struct?

Reply:
There was some reason I don't remember to use `uint64_t`. But I think using a byte array in the struct is arbitrary... probably easier to read if replaced by `uint64_t`.

Reply:
Another thing, I'm pretty sure memcpy from `uint64_t`  to `unsigned char[8]` is platform defined(BE/LE)

Reply:
Yes, but not really an issue as the doc clearly states

> The exact representation of data inside is implementation defined and not guaranteed to be portable between different platforms or versions.

Reply:
It doesn't matter in the end but the code is certainly easier to read if it's the same type twice (no matter what type that is)